src/ray/raylet/agent_manager.cc

```diff
@@ -80,34 +86,34 @@ void AgentManager::StartAgent() {
   ProcessEnvironment env;
   env.insert({"RAY_NODE_ID", options_.node_id.Hex()});
   env.insert({"RAY_RAYLET_PID", std::to_string(getpid())});
-  Process child(argv.data(), nullptr, ec, false, env);
-  if (!child.IsValid() || ec) {
+  agent_process_ = std::make_unique<Process>(argv.data(), nullptr, ec, false, env);
+  if (!agent_process_->IsValid() || ec) {
     // The worker failed to start. This is a fatal error.
     RAY_LOG(FATAL) << "Failed to start agent with return value " << ec << ": "
                    << ec.message();
   }
 
-  std::thread monitor_thread([this, child]() mutable {
+  std::thread monitor_thread([this]() mutable {
     SetThreadName("agent.monitor");
-    RAY_LOG(INFO) << "Monitor agent process with pid " << child.GetId()
+    RAY_LOG(INFO) << "Monitor agent process with pid " << agent_process_->GetId()
                   << ", register timeout "
                   << RayConfig::instance().agent_register_timeout_ms() << "ms.";
     auto timer = delay_executor_(
-        [this, child]() mutable {
-          if (agent_pid_ != child.GetId()) {
-            RAY_LOG(WARNING) << "Agent process with pid " << child.GetId()
-                             << " has not registered. ip " << agent_ip_address_
-                             << ", pid " << agent_pid_;
-            child.Kill();
+        [this]() mutable {
+          if (!IsAgentRegistered()) {
+            RAY_LOG(WARNING) << "Agent process with pid " << agent_process_->GetId()
+                             << " has not registered. ip " << agent_info_.ip_address()
+                             << ", pid " << agent_info_.pid();
+            agent_process_->Kill();
```

Comment:
Question about the lifecycle. I heard that agent is fate shared with raylet, but here it doesn't crash the raylet.

Reply:
@iycheng In here: https://github.com/ray-project/ray/blob/1949f359014b8be69709110772a54fc2f0c1cab8/src/ray/raylet/agent_manager.cc#L106

agent_manager will wait for the agent to exit and kill the current raylet process by `QuickExit`

Reply:
Got it. Thanks for the explanation!